sequence_processing_pipeline/NuQCJob.py

```diff
@@ -0,0 +1,291 @@
+
+            if not exists(filtered_directory):
+                raise PipelineError(f"{filtered_directory} does not exist")
+
+            empty_files_directory = join(source_dir, 'zero_files')
+            self._filter_empty_fastq_files(filtered_directory,
+                                           empty_files_directory,
+                                           self.minimum_bytes)
+
+    def _get_failed_indexes(self, project_name, job_id):
+        pattern = f"{self.temp_dir}/{self.batch_prefix}.*.completed"
+        completed_files = list(glob.glob(pattern))
+        completed_indexes = []
+        regex = r'%s.%s_([0-9]+).completed' % (self.batch_prefix, str(job_id))
+        array_ids = re.compile(regex)
+
+        for completed_file in completed_files:
+            match = array_ids.search(completed_file)
+            if match is None:
+                raise PipelineError("Malformed completed file")
+            else:
+                id_ = int(match.groups(0)[0])
+                completed_indexes.append(id_)
+
+        # a successfully completed job array should have a list of array
+        # numbers from 0 - len(self.commands).
+        all_indexes = list(range(1, self.counts[self.batch_prefix]))
+        failed_indexes = sorted(set(all_indexes) - set(completed_indexes))
+
+        # generate log-file here instead of in run() where it can be
+        # unittested more easily.
+        log_fp = join(self.output_path,
+                      'logs',
+                      f'failed_indexes_{job_id}.json')
+
+        if failed_indexes:
+            with open(log_fp, 'w') as f:
+                f.write(dumps({'job_id': job_id,
+                               'failed_indexes': failed_indexes}, indent=2))
+
+        return failed_indexes
+
+    def _process_sample_sheet(self):
+        sheet = KLSampleSheet(self.sample_sheet_path)
+        valid_sheet = validate_and_scrub_sample_sheet(sheet)
+
+        if not valid_sheet:
+            s = "Sample sheet %s is not valid." % self.sample_sheet_path
+            raise PipelineError(s)
+
+        header = valid_sheet.Header
+        chemistry = header['chemistry']
+
+        if header['Assay'] not in Pipeline.assay_types:
+            s = "Assay value '%s' is not recognized." % header['Assay']
+            raise PipelineError(s)
+
+        sample_ids = []
+        for sample in valid_sheet.samples:
+            sample_ids.append((sample['Sample_ID'], sample['Sample_Project']))
+
+        bioinformatics = valid_sheet.Bioinformatics
+
+        # reorganize the data into a list of dictionaries, one for each row.
+        # the ordering of the rows will be preserved in the order of the list.
+        lst = bioinformatics.to_dict('records')
+
+        # convert true/false and yes/no strings to true boolean values.
+        for record in lst:
+            for key in record:
+                if record[key].strip().lower() in ['true', 'yes']:
```

Comment:
If some information is manually provided unexpectedly, should we instead raise a exception on the assumption that other data may be bad?

Reply:
Based on our interactions with the wet-lab, the rigor would be nice for us but it would quickly become odious for them; it would lead to a lot of extra failed submit-jobs and back-and-forth. It's better to continue to rely on catching and reporting specific errors back to them in a helpful way.

Reply:
Please do raise an exception if we receive unexpected data in a sample sheet. I checked with Rob out-of-band that this is the correct action. 

cc @rob-knight